File: Problemset/1052-Grumpy Bookstore Owner/Sliding Window.rb
# @param {Integer[]} customers
# @param {Integer[]} grumpy
# @param {Integer} x
# @return {Integer}
def max_satisfied(customers, grumpy, x)
    curr = 0
    ret = 0

    for i in 0...grumpy.length
        if i < x or grumpy[i] == 0
            curr += customers[i]
            ret += customers[i]
        end
        curr -= customers[i - x] if i >= x and grumpy[i - x] == 1
        curr += customers[i] if i >= x and grumpy[i] == 1

        ret = [ret, curr].max
    end

    return ret
end

File: Problemset/1052-Grumpy Bookstore Owner/Sliding Window.rs
impl Solution {
    pub fn max_satisfied(customers: Vec<i32>, grumpy: Vec<i32>, x: i32) -> i32 {
        let mut curr = 0;
        let mut ret = 0;

        for i in 0..grumpy.len() {
            if i < x as usize || grumpy[i] == 0 {
                curr += customers[i];
                ret += customers[i];
            }
            if i >= x as usize && grumpy[i - x as usize] == 1 {
                curr -= customers[i - x as usize];
            }
            if i >= x as usize && grumpy[i] == 1 {
                curr += customers[i];
            }

            ret = ret.max(curr);
        }

        ret
    }
}
